library/alloc/src/string.rs

```diff
@@ -3094,6 +3101,92 @@ impl fmt::Write for String {
     }
 }
 
+/// Placeholder docs.
+#[unstable(feature = "into_chars", reason = "new API", issue="0")]
+pub struct IntoChars {
+    bytes: Vec<u8>,
+}
+
+#[unstable(feature = "into_chars", reason = "new API", issue="0")]
+impl fmt::Debug for IntoChars {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("IntoChars").field(&self.as_str()).finish()
+    }
+}
+
+#[unstable(feature = "into_chars", reason = "new API", issue="0")]
+impl IntoChars {
+    /// Placeholder docs.
+    pub fn as_str(&self) -> &str {
+        // SAFETY: `bytes` is a valid UTF-8 string.
+        unsafe { str::from_utf8_unchecked(self.bytes.as_slice()) }
+    }
+
+    fn iter(&self) -> CharIndices {
+        self.as_str().char_indices()
+    }
+}
+
+#[unstable(feature = "into_chars", reason = "new API", issue="0")]
+impl AsRef<str> for IntoChars {
+    fn as_ref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+#[unstable(feature = "into_chars", reason = "new API", issue="0")]
+impl AsRef<[u8]> for IntoChars {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes.as_slice()
+    }
+}
+
+#[unstable(feature = "into_chars", reason = "new API", issue="0")]
+impl Iterator for IntoChars {
+    type Item = char;
+
+    #[inline]
+    fn next(&mut self) -> Option<char> {
+        let mut iter = self.iter();
+        match iter.next() {
+            None => None,
+            Some((_, ch)) => {
+                let offset = iter.offset();
+                drop(self.bytes.drain(..offset));
```

Comment:
This reminds me of [this thread](https://users.rust-lang.org/t/best-way-to-drop-range-of-elements-from-front-of-vecdeque/31795/4) that suggests a `truncate_front` method on `Vec`. Not sure if it's desired as a separate proposal.